corehq/apps/enterprise/iterators.py

```diff
@@ -1,7 +1,227 @@
+
+        it = loop_over_domains(
+            domains,
+            MobileFormSubmissionsQueryFactory(),
+            limit=limit,
+            last_domain=self.last_domain,
+            start_date=self.start_date,
+            end_date=self.end_date,
+            last_time=self.last_time,
+            last_id=self.last_id,
+        )
+
+        xform_converter = RawFormConverter()
+        return (xform_converter.convert(form) for form in it)
+
+    @classmethod
+    def get_kwargs_from_map(cls, map):
+        last_domain = map.get('domain', None)
+        last_time = map.get('inserted_at', None)
+        if last_time:
+            last_time = datetime.fromisoformat(last_time).astimezone(timezone.utc)
+        last_id = map.get('id', None)
+        return {
+            'last_domain': last_domain,
+            'last_time': last_time,
+            'last_id': last_id
+        }
+
+    @classmethod
+    def get_query_params(cls, fetched_object):
+        return {
+            'domain': fetched_object['domain'],
+            'inserted_at': fetched_object['inserted_at'],
+            'id': fetched_object['form_id']
+        }
+
+
+def resolve_start_and_end_date(start_date, end_date, maximum_date_range):
+    '''
+    Provide start and end date values if not supplied.
+    '''
+    if not end_date:
+        end_date = datetime.now(timezone.utc)
+
+    if not start_date:
+        start_date = end_date - timedelta(days=30)
+
+    if end_date - start_date > timedelta(days=maximum_date_range):
+        raise TooMuchRequestedDataError(
+            _('Date ranges with more than {} days are not supported').format(maximum_date_range)
+        )
+
+    return start_date, end_date
+
+
+class RawFormConverter:
+    def __init__(self):
+        self.app_lookup = AppIdToNameResolver()
+
+    def convert(self, form):
+        domain = form['domain']
+        submitted_date = datetime.strptime(form['received_on'][:19], '%Y-%m-%dT%H:%M:%S')
+        inserted_at = datetime.strptime(form['inserted_at'][:19], '%Y-%m-%dT%H:%M:%S')
+
+        return {
+            'form_id': form['form']['meta']['instanceID'],
+            'form_name': form['form']['@name'] or _('Unnamed'),
+            'submitted': submitted_date,
+            'inserted_at': inserted_at,
+            'app_name': self.app_lookup.resolve_app_id_to_name(domain, form['app_id']) or _('App not found'),
+            'username': form['form']['meta']['username'],
+            'domain': domain
+        }
+
+
+class AppIdToNameResolver:
+    def __init__(self):
+        self.domain_lookup_tables = {}
+
+    def resolve_app_id_to_name(self, domain, app_id):
+        if 'domain' not in self.domain_lookup_tables:
+            domain_apps = get_brief_apps_in_domain(domain)
+            self.domain_lookup_tables[domain] = {a.id: a.name for a in domain_apps}
+
+        return self.domain_lookup_tables[domain].get(app_id, None)
+
+
+def loop_over_domains(domains, query_factory, limit=None, last_domain=None, **kwargs):
```

Comment:
maybe call this `run_query_over_domains`? and maybe switch the order of `query_factory` and `domains`